### PR TITLE
Fix fmod and remainder returning NaN on the vectorized CPU path

### LINKED ISSUE #193753: torch.fmod and torch.remainder return NaN on the vectorized path where the scalar path is correct (the fmod half of #77742)
### 🐛 Describe the bug

`torch.fmod` and `torch.remainder` return NaN when `|a / b|` overflows, but only for tensors long enough to reach the vectorized kernel. The same values in a shorter tensor give the correct answer.

```python
import torch, math

for n in (3, 4):
    x = torch.full((n,), 1e308, dtype=torch.float64)
    y = torch.full((n,), 0.5, dtype=torch.float64)
    print(n, torch.fmod(x, y)[0].item(), torch.remainder(x, y)[0].item())
# 3 0.0 0.0
# 4 nan nan

print(math.fmod(1e308, 0.5))  # 0.0
```

float32 is the same, with the threshold at 8 elements:

```python
x = torch.full((8,), 3.4028235e38, dtype=torch.float32)
y = torch.full((8,), -1.4e-45, dtype=torch.float32)
print(torch.remainder(x, y)[0].item())  # nan
print(torch.remainder(x[:1], y[:1])[0].item())  # 0.0
# numpy and Python both give -0.0
```

This is the case @peterbell10 posted in #77742. That issue was closed by #129536, which fixed `div` with `rounding_mode="floor"` by working around the bad `fmod` result inside `div_floor_floating_vec`:

https://github.com/pytorch/pytorch/blob/04b971abedf3167a3f929e4e7a9a5b7e8647c02c/aten/src/ATen/native/cpu/BinaryOpsKernel.cpp#L261-L263

`fmod_kernel` and `remainder_kernel` call `Vectorized::fmod` with no such fixup, so the `torch.fmod` half of #77742 is still there:

- `fmod_kernel`: https://github.com/pytorch/pytorch/blob/04b971abedf3167a3f929e4e7a9a5b7e8647c02c/aten/src/ATen/native/cpu/BinaryOpsKernel.cpp#L1056
- `remainder_kernel`: https://github.com/pytorch/pytorch/blob/04b971abedf3167a3f929e4e7a9a5b7e8647c02c/aten/src/ATen/native/cpu/BinaryOpsKernel.cpp#L404
- `remainder_kernel`, bf16 path: https://github.com/pytorch/pytorch/blob/04b971abedf3167a3f929e4e7a9a5b7e8647c02c/aten/src/ATen/native/cpu/BinaryOpsKernel.cpp#L382

The cause is the one already identified in #77742: `Sleef_fmod` is documented as undefined once `|x / y|` reaches 1e300 for double or 1e38 for float. Every vec backend implements `Vectorized<T>::fmod` as a direct Sleef call, so this is not specific to one ISA.

I know results are not guaranteed bitwise identical across vectorization. This is not a rounding difference. It is NaN in place of a finite value, and it flips based on tensor length alone.

One thing to watch in a fix: the result is not always zero when the division overflows, so the value cannot just be replaced with 0.

```python
math.fmod(-1.7976931348623157e308, 1.5e-323)  # -1e-323
```

Using the scalar `std::fmod` for the lanes where `(a / b).abs()` is infinite and `a` is finite would handle it, and that mask is already being computed in `div_floor_floating_vec`. Glad to send a PR if that is the direction you want.

Found by a differential tester that varies tensor length while holding the values fixed.

### Versions

torch 2.13.0, CPU, macOS arm64.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @aditew01

### PR DESCRIPTION
**Pull request by krrishapatel**
Fixes #193753.

`torch.fmod` and `torch.remainder` return `NaN` on the vectorized CPU path where the scalar path is correct. No extremal divisor is needed, just a large dividend:

```python
>>> import torch
>>> a = torch.full((64,), 1e38)
>>> torch.fmod(a, 1e-3)[0]        # vectorized
tensor(nan)
>>> torch.fmod(a[0], 1e-3)        # scalar, same inputs
tensor(9.2320e-05)
```

### Cause

`Vectorized<T>::fmod` calls `Sleef_fmod`, and Sleef documents its result as undefined once `abs(a / b)` gets large. The mechanism is in `xfmod`: the reduction step forms `trunc(r * (1 / abs(y)))`, that product overflows to infinity, and the double-double update then turns the accumulator into `NaN`. So the failure signature is always a `NaN`.

@peterbell10 diagnosed this on #77742 in 2022 and posted the `fmod` case there. #129536 closed that issue, but it only fixed `div` with `rounding_mode="floor"`, and it did so by throwing the remainder away:

```cpp
auto floor = vec_t::blendv(a - mod, a, (basic_div.abs() == inf) & (a.abs() != inf));
```

`div_floor` can do that because it only needs the quotient. For `fmod` and `remainder` the remainder is the result, so both still call Sleef unguarded.

### Fix

`fmod_floating_vec` recomputes the vector with `std::fmod` when `Vectorized<T>::fmod` returns a `NaN` lane.

Detecting the `NaN` in the output is cheaper than predicting overflow from the inputs, and it does not depend on knowing where Sleef stops being accurate. It is deliberately a superset: a legitimate `NaN`, from a `NaN` input or a zero divisor, takes the scalar path too and still comes back `NaN`.

Keeping the fallback out of line matters. Inlined into the vectorized loop it cost about 2x even on inputs that never reach it, which I measured in isolation against the vec headers before touching the kernel.

Call sites changed: `fmod_kernel`, the generic `remainder_kernel`, and both halves of the bfloat16 `remainder` specialization. The `a.fmod(b)` inside `div_floor_floating_vec` is left alone since #129536 already covers it.

float and double are guarded. `Vectorized<Half>` and `Vectorized<BFloat16>` implement `fmod` with `std::fmod` per element, so they are already exact. bfloat16 `remainder` still needs the fix because that kernel converts to float first, and bfloat16 has float32's exponent range:

```python
>>> b = torch.full((32,), 2.37e38, dtype=torch.bfloat16)
>>> torch.remainder(b, 0.5)[0]    # before: nan, scalar path gives 0
```

### Correctness

32 combinations of `+/- max * 0.7` over `+/- tiny` and `+/- 0.5`, in float32 and float64, at shape `()` and `(32,)`, for both ops:

| | before | after |
|---|---|---|
| mismatches vs NumPy | 32 | 0 |

Every failure was on the length-32 tensors, which is what pins it to the vectorized kernel.

### Performance

Single threaded, arm64, best of 7, uniform inputs that never trigger the fallback. Run-to-run spread on this machine is about 2.5%.

| n = 2^20 | before | after | |
|---|---|---|---|
| `fmod` float32 | 1278 us | 1421 us | +11% |
| `remainder` float32 | 1454 us | 1586 us | +9% |
| `fmod` float64 | 2686 us | 2955 us | +10% |
| `remainder` float64 | 3041 us | 3253 us | +7% |

So about 10%, and it is the branch, not the check. Two things I measured and rejected: predicting the overflow from the inputs with a division instead, which came out the same or slightly worse and needs a threshold; and a branchless version, which needs two Sleef calls per vector and is far worse. Some of the remaining cost is `Vectorized<double>::zero_mask()` on NEON extracting lanes one at a time where the float version uses a horizontal add. That looks like a one-line improvement, but it is a separate change so I left it out.

Happy to take this a different way if you would rather not pay 10% here.

### Tests

`test_fmod_remainder_extremal` and `test_remainder_extremal_bfloat16` in `test/test_binary_ufuncs.py`, next to the existing `test_fmod_remainder_overflow`. Each runs at shape `()` and `(32,)`, so before this change they fail on the vectorized path and pass on the scalar one. The float reference is NumPy applied to the tensors the kernel actually saw, not to the Python floats, since `fmod` of a ratio this large moves with one ULP of rounding in the cast. The bfloat16 case uses the scalar path as the reference.

`python -m pytest test/test_binary_ufuncs.py` passes.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @aditew01

### CODE REVIEW

**Comment by albanD** [comment]
Thank you very much for sending a PR.
To avoid you spending time preparing a PR for a feature we're not sure we want to add, our usual process is to discuss on the issue before doing the PR.
In particular, make sure that the issue is marked "actionable" before sending a PR!
See https://github.com/pytorch/pytorch/wiki/The-Ultimate-Guide-to-PyTorch-Contributions for details!
